File: enigma/JournalService.py
class JournalEntry:
    pass

class JournalException(Exception):

    def __init__(self, id, label):

        self.journalEntry = JournalEntry()
        self.journalEntry.id = id
        self.journalEntry.label = label

    def __init__(self, journalEntry):

        self.journalEntry = JournalEntry()
        self.journalEntry.id = journalEntry.id
        self.journalEntry.label = journalEntry.label

    def __str__(self):
        return repr(self.journalEntry)

journalInstance = []

def journal_add(id, label):

    global journalInstance

    journalEntry = JournalEntry()
    journalEntry.id = id
    journalEntry.label = label

    journalInstance.append(journalEntry)

    raise JournalException(journalEntry)
    return journalEntry

def journal_get(index = -1):

    global journalInstance

    if not journalInstance:
        return None

    if index < len(journalInstance):
        return journalInstance[index];

    return None

def journal_pop():

    global journalInstance

    if not journalInstance:
        return None

    return journalInstance.pop()

File: js/JournalService.js
function JournalEntry(id, label) {

    this.id = id;
    this.label = label; 

    return; }

function journal_entry_set(journalEntry, id, label) {

    journalEntry.id = id;
    journalEntry.label = label;

    return journalEntry; }

function JournalException(journalEntry) {

    this.journalEntry = journalEntry;
    return; }

journalInstance = new Array();

function journal_add(id, label) {

    journalEntry = new JournalEntry(id, label);
    journalInstance.push(new JournalEntry(id, label));

    throw new JournalException(journalEntry);
    return journalEntry; }

function journal_get(id) {

    id = (typeof id == 'undefined') ?
        journalInstance.length-1: id;

    if (id < journalInstance.length)
        return journalInstance[id];

    return null; }

function journal_pop() {

    if (journalInstance.length == 0)
        return null;

    return journalInstance.pop(); }
